app/services/workflow_create_service.rb

```diff
@@ -14,7 +14,7 @@ def create(options)
     begin
       retries ||= 0
       template.workflows.create!(options)
-    rescue ActiveRecord::RecordNotUnique # The auto generated sequence number may be found duplicated due to concurrent issue
+    rescue ActiveRecord::RecordNotUnique, Exceptions::NegativeSequenceError # The auto generated sequence number may be found duplicated due to concurrent issue
       retry if (retries += 1) < 3
```

Comment:
@bzwei I get that the model will raise the new `NegativeSequenceError` message on create, update or destroy but how does retrying potentially fix the situation?

And when we run out of retries it means we need a manual process to correct it?  Did the gem we were using before have this same exposure?

Reply:
True. It is not necessary to catch `NegativeSequenceError` for create. I added to make the same as how we catch similar errors in other places. I can remove it.

Reply:
My question is really more about the logic in general.

1. How does the retry fix the issue?
2. How do we fix this when we run out of retires and the issue still exists?
3. Did the gem have the same exposure?

Reply:
The gem we used before did not allow to make `sequence` unique. So it does raise an error but may result in bad record.
If retry does not solve the problem (should be very rare), then yes, we will fail the api so the user can try later. But we should not need a manual fix in the database.

Reply:
The above comment applies to update and destroy operations. For creation there is no reload. But retry helps to find newest available sequence thus avoid duplicates.